File: ex20.go
func isValid(s string) bool {
    if s == "" {
        return true
    }
    var stack []string
    for i := 0; i < len(s); i++ {
        if s[i] == '(' || s[i] == '{' || s[i] == '[' {
            stack = append(stack, string(s[i]))
        } else {
            if len(stack) <= 0 {
                return false
            }
            char := stack[len(stack)-1]
            stack = stack[:len(stack)-1]
            if (char == "(" && s[i] != ')') || (char == "{" && s[i] != '}') || (char == "[" && s[i] != ']') {
                return false
            }
        }
    }
    return len(stack) == 0
}

File: leetcode/ex20.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == ')' || s[i] == '}' || s[i] == ']') {
                if (st.empty())
                    return false;
                char value = st.top();
                st.pop();
                if (value != '(' && s[i] == ')') return false;
                if (value != '{' && s[i] == '}') return false;
                if (value != '[' && s[i] == ']') return false;
            } else {
                st.push(s[i]);
            }
        }
        return st.empty();
    }
};
